### PR TITLE
token balances on token listing for address

### LINKED ISSUE #564: Contributor sees the Token balance for each Token listed in the Token tab for the Address page
As a `Contributor`, I want to see a `Token` tab in the `Address Details` page, where I can see the list of `Tokens` the `Address` holds, as well as see the `Token Transfers` for each of those `Tokens`.

https://projects.invisionapp.com/share/8EMX9IOR2V9#/screens/308371747

This continues the work on #410.

## Acceptance Criteria

- For each `Token`, there should be possible to see the following information:
    - `Token` balance (for the associated address) followed by the token symbol.
        - For `ERC20 Tokens`, the balance should be formatted as currency (thousands and decimal separators) considering the token's `decimals` info (if there is no `decimals` info, display the formatted amount). Example: `999,000.02 FREH`
        - For `ERC721 Tokens`, the balance of non-fungible tokens. Example: `803,843 CK`
        - In case the `Token` has no `symbol`, show nothing next to the amount.
    -  If `balance = 0`, the `Token` should be at the bottom of the list.

### Out of Scope
- This issue doesn't include the search field as shown in the mockup.
- This issue doesn't include the token icons as shown in the mockup.
- This issue doesn't include the token balance conversion to USD, which includes:
  - the USD exchange rate for that specific token
  - how much USD that amount of token is worth

### Dependencies
- [x] Depends on #514.

### Task breakdown
- [x] Change query that gets the tokens an address holds to include the token balance.
  - [x] Considers the last block.
- [x] Make the balance available in the Token struct, like the number of transfers.*
- [x] Update the front-end to include the balance.
  - [x] Format just Tokens that have the type ERC-20 (use `CurrencyHelper`)
  - [x] Consider symbol at the end of the value. Ex. (100 Freh)

### PR DESCRIPTION
**Pull request by GuSilverFlame**
resolves #564
### Enhancements

- bring token balances along with the count of transfers

- change the query to do all the needed operations at once 

### Incompatible Changes

- removed from Explorer.Chain the function count_token_transfers_from_address_hash/2
  reason: it was unused in the project after the changes

- removed from Explorer.Chain the function fetch_tokens_from_address_hash/2
  reason: it was unused in the project after the changes

- removed from Explorer.Chain.Token the function interactions_with_address/2
  reason: it was unused in the project after the changes

- the function with_transfers_by_address/2 was changed and no longer returns a list of %Token{}
  reason: group_by requires that all the fields to be selected are specified and some  of the needed fields don't belong to Token

## Upgrading

all the removed functions are simple and easy to reimplement should you need any of them,
they were removed just for being unused after the changes

## screen capture
captura de tela 2018-09-19 as 17 00 32

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex [hunk 1]
```diff
@@ -1,10 +1,15 @@
 <div class="tile tile-type-token">
-  <div class="row justify-content">
-    <div class="col-md-12 d-flex flex-column tile-label">
+  <div class="row justify-content align-items-center">
+    <div class="col-md-8 d-flex flex-column tile-label">
       <%= link(to: token_path(@conn, :show, @token.contract_address_hash), class: "tile-title-lg") do %>
-        <%= token_name(@token) %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex]: Notice that `token_name` is a [helper](https://github.com/poanetwork/blockscout/blob/gsf-token-balances-on-token-listing-for-address/apps/block_scout_web/lib/block_scout_web/views/tokens/helpers.ex#L61).

We need to keep the same behavior when the `name` is nil and probably `blank`. We can use the same helper adding a new function or just removing the `struct Token` to pattern match against a `map` instead 🤔.

File: apps/block_scout_web/test/block_scout_web/controllers/address_token_controller_test.exs [hunk 2]
```diff
@@ -89,6 +119,13 @@ defmodule BlockScoutWeb.AddressTokenControllerTest do
 
       Enum.each(1..51, fn i ->
         token = insert(:token, name: "A Token#{i}", type: "ERC-20")
+
+        insert(:token_balance,
```
Comment (feliperenan) [inline apps/block_scout_web/test/block_scout_web/controllers/address_token_controller_test.exs]: 🎨 we can use the same indentation that you're using above ☝️.

File: apps/explorer/lib/explorer/chain/token.ex [hunk 3]
```diff
@@ -90,44 +91,54 @@ defmodule Explorer.Chain.Token do
...
-        token in Token,
-        join: tt in TokenTransfer,
-        on: tt.token_contract_address_hash == token.contract_address_hash,
-        where: tt.to_address_hash == ^address_hash or tt.from_address_hash == ^address_hash,
-        distinct: [:contract_address_hash]
+        tb in TokenBalance,
+        where: tb.address_hash == ^address_hash,
+        distinct: :token_contract_address_hash,
+        order_by: [desc: :block_number],
+        select: %{value: tb.value, token_contract_address_hash: tb.token_contract_address_hash}
```
Comment (feliperenan) [inline apps/explorer/lib/explorer/chain/token.ex]: Do we also need to specify the fields here or is it for optimization?

File: apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
+        <%= @token.name %>
       <% end %>
-      <span><%= @token.type %> - <%= number_of_transfers(@token) %></span>
```
Comment (amandasposito) [inline apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex]: Why did you remove the helper from here?

File: apps/explorer/lib/explorer/chain/token.ex (hunk 3 above)
Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/token.ex]: the fields are specified here just for optimization

File: apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex (hunk 4 above)
Comment (GuSilverFlame) [reply apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex]: good question, I don't know why, might put it back

File: apps/block_scout_web/test/block_scout_web/controllers/address_token_controller_test.exs (hunk 2 above)
Comment (GuSilverFlame) [reply apps/block_scout_web/test/block_scout_web/controllers/address_token_controller_test.exs]: will do, this one probably got like this when I ran the formatter

File: apps/explorer/test/explorer/chain_test.exs [hunk 5]
```diff
@@ -2306,228 +2313,89 @@ defmodule Explorer.ChainTest do
...
-    test "orders by type, name and inserted_at time" do
-      address = insert(:address)
-
-      first_token =
-        :token
-        |> insert(name: "token-c", type: "ERC-721")
-        |> Repo.preload(:contract_address)
-
-      second_token =
+      other_token =
```
Comment (feliperenan) [inline apps/explorer/test/explorer/chain_test.exs]: IMO we can remove this one from the test setup since we are testing if tokens with zero balance will not be considered. If we want to test if the correct token will be returned, we can create another test for it. Wdyt?

File: apps/explorer/test/explorer/chain_test.exs [hunk 6]
```diff
@@ -2306,228 +2313,89 @@ defmodule Explorer.ChainTest do
...
-        |> Enum.map(&Repo.preload(&1, :contract_address))
+        |> Chain.tokens_with_number_of_transfers_from_address()
+        |> Enum.find(fn t -> t.name == "atoken" end)
 
-      assert fetched_tokens == [first_token, second_token, third_token, fourth_token]
+      assert fetched_token == nil
     end
 
-    test "supports pagination" do
+    test "brings the value of the last balance" do
```
Comment (feliperenan) [inline apps/explorer/test/explorer/chain_test.exs]: I would add a test to check if the query always will consider the last block even if the token balance is `0`. In this case, the query should skip this one from the result.

File: apps/explorer/lib/explorer/chain/token.ex [hunk 7]
```diff
@@ -90,44 +91,54 @@ defmodule Explorer.Chain.Token do
...
       from(
-        token in Token,
-        join: tt in TokenTransfer,
-        on: tt.token_contract_address_hash == token.contract_address_hash,
-        where: tt.to_address_hash == ^address_hash or tt.from_address_hash == ^address_hash,
-        distinct: [:contract_address_hash]
+        tb in TokenBalance,
+        where: tb.address_hash == ^address_hash,
+        distinct: :token_contract_address_hash,
+        order_by: [desc: :block_number],
```
Comment (amandasposito) [inline apps/explorer/lib/explorer/chain/token.ex]: is this necessary?

File: apps/explorer/test/explorer/chain_test.exs [hunk 8]
```diff
@@ -2542,18 +2410,12 @@ defmodule Explorer.ChainTest do
...
-    end
-
-    test "returns 0 if no interaction is found" do
-      address = insert(:address)
-
-      token =
-        :token
-        |> insert(name: "token")
-        |> Repo.preload(:contract_address)
+      fetched_token =
```
Comment (feliperenan) [inline apps/explorer/test/explorer/chain_test.exs]: Since we are testing the last balance, we can remove `token_transfer` factories above.

File: apps/explorer/lib/explorer/chain/token.ex (hunk 7 above)
Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/token.ex]: this is to make sure only the last balance comes

File: apps/explorer/test/explorer/chain_test.exs [hunk 9]
```diff
@@ -2306,228 +2313,89 @@ defmodule Explorer.ChainTest do
...
-
-      third_token =
-        :token
-        |> insert(name: "token-b", type: "ERC-20")
-        |> Repo.preload(:contract_address)
-
-      fourth_token =
-        :token
-        |> insert(name: "token-b", type: "ERC-20", inserted_at: third_token.inserted_at)
+        |> insert(name: "token-c", type: "ERC-721", decimals: 0, symbol: "TC")
```
[the remark this answers is not in the document]
Comment (feliperenan) [reply apps/explorer/test/explorer/chain_test.exs]: Nevermind.

File: apps/explorer/test/explorer/chain_test.exs (hunk 8 above)
Comment (GuSilverFlame) [reply apps/explorer/test/explorer/chain_test.exs]: removed 👍

File: apps/explorer/lib/explorer/chain/token.ex [hunk 10]
```diff
@@ -90,44 +91,54 @@ defmodule Explorer.Chain.Token do
...
     from(
       t in Token,
       join: tt in TokenTransfer,
       on: tt.token_contract_address_hash == t.contract_address_hash,
-      where: t.contract_address_hash == ^token_hash,
-      where: tt.to_address_hash == ^address_hash or tt.from_address_hash == ^address_hash,
-      select: tt
+      join: tb in subquery(last_balance_query),
+      on: tb.token_contract_address_hash == t.contract_address_hash,
+      order_by: [desc: t.type, asc: fragment("UPPER(?)", t.name)],
```
Comment (amandasposito) [inline apps/explorer/lib/explorer/chain/token.ex]: what is the difference in ordering by with upper case? 🤔

File: apps/block_scout_web/lib/block_scout_web/chain.ex [hunk 11]
```diff
@@ -185,7 +184,7 @@ defmodule BlockScoutWeb.Chain do
     %{"inserted_at" => inserted_at_datetime}
   end
 
-  defp paging_params(%Token{name: name, type: type, inserted_at: inserted_at}) do
+  defp paging_params(%{name: name, type: type, inserted_at: inserted_at}) do
```
Comment (igorffs) [inline apps/block_scout_web/lib/block_scout_web/chain.ex]: These functions `paging_params` and `token_name` are Token specific, why have we removed the matching on the Struct?

File: apps/explorer/lib/explorer/chain/token.ex [hunk 12]
```diff
@@ -90,44 +91,54 @@ defmodule Explorer.Chain.Token do
   def with_transfers_by_address(address_hash, options \\ []) do
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/token.ex]: Can we add tests for this function? \o

File: apps/explorer/lib/explorer/chain/token.ex [hunk 13]
```diff
@@ -90,44 +91,54 @@ defmodule Explorer.Chain.Token do
...
+        decimals: max(t.decimals),
+        type: t.type,
+        transfers: count(t.name)
+      }
     )
   end
 
   def page_token(query, %PagingOptions{key: nil}), do: query
 
   def page_token(query, %PagingOptions{key: {name, type, inserted_at}}) do
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/token.ex]: It'd be nice to test this function too. It'd also be nice including some docs since it isn't something so straightforward anymore. wdyt?

**Comment by acravenho** [comment]
@GuSilverFlame It's looking good. I think we should use an identical layout as #744. Push the balance to the right and the percentage under the balance. They are also including the count 1-250 and removing pagination.

File: apps/explorer/lib/explorer/chain/token.ex (hunk 12 above)
Comment (igorffs) [reply apps/explorer/lib/explorer/chain/token.ex]: It'd be awsome if you could include some docs in this function too. 🙏

Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/token.ex]: we can without a problem but it's pretty much all tested in apps/explorer/test/explorer/chain_test.exs where the function 'tokens_with_number_of_transfers_from_address/2' is being tested
is there some benefit in testing just the query without executing it? if so it'd be a type of test I could specifically for this function

File: apps/explorer/lib/explorer/chain/token.ex (hunk 10 above)
Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/token.ex]: it's because the ordering on the database seems to be case sensitive, when I was testing with the address `0x0000000000000000000000000000000000000000` the token 
`aXpire Token` was showing up *after* `Wrapped Ether` before I did this upper case change

File: apps/explorer/lib/explorer/chain/token.ex (hunk 7 above)
Comment (amandasposito) [reply apps/explorer/lib/explorer/chain/token.ex]: But you are not limiting the results here, you are only ordering, so if a `contract_address_hash` has more than one register in this table, it would bring all the records in this subquery, this would duplicate the tokens lists. 

Can we write a test for this scenario?

File: apps/explorer/lib/explorer/chain/token.ex (hunk 12 above)
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/token.ex]: We should probably rename this function due to its change in behavior. Only "with_transfers" isn't expressing the whole thing anymore. The most probable name here would be the same name we have in the private function "with_last_balance_and_transfers_count". 🤔 

This name ☝️ sounds an alarm ⏰ for me, we might be accumulating too many responsibilities in this single function. Do you think it'd be possible to create different queries and make them be composable that could allow something like this:

```elixir
Token
|> with_token_transfer_count
|> with_token_balance
```

File: apps/explorer/lib/explorer/chain/token.ex (hunk 7 above)
Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/token.ex]: I'm limiting it with the distinct clause

File: apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex [hunk 14]
```diff
@@ -1,10 +1,15 @@
 <div class="tile tile-type-token">
-  <div class="row justify-content">
-    <div class="col-md-12 d-flex flex-column tile-label">
+  <div class="row justify-content align-items-center">
+    <div class="col-md-8 d-flex flex-column tile-label">
```
Comment (arthurra) [inline apps/block_scout_web/lib/block_scout_web/templates/address_token/_tokens.html.eex]: Adding some white space here on small screens will help to break up the content within the tile and make it easier to read. You can use Bootstrap helpers to control this based on screen size: 

the classes `mb-3 mb-md-0` will add white space on phone-sized screens and remove on tablets and desktop.

File: apps/explorer/lib/explorer/chain/token.ex (hunk 7 above)
Comment (amandasposito) [reply apps/explorer/lib/explorer/chain/token.ex]: Oh, I got it. Postgres supports `distinct on`, so the `distinct` here with the `order by` won't let the results to be duplicated. That's nice 👍 

I'm okay with going on with only the `chain_test.exs` 👍

File: apps/block_scout_web/lib/block_scout_web/chain.ex (hunk 11 above)
Comment (amandasposito) [reply apps/block_scout_web/lib/block_scout_web/chain.ex]: It is a question to decide if removing the `struct` from the pattern matching will generate some unexpected behavior with these functions if so we can return a `struct` from the `ecto select` as well and solve this, probably not the `Explorer.Chain.Token` because it doesn't have the `value` and `transfers` fields in int.

Comment (GuSilverFlame) [reply apps/block_scout_web/lib/block_scout_web/chain.ex]: as long as all the keys used by the function are in the pattern matching, the behavior will remain the same as if it was a Token struct, but I could create a special struct for this case there's no problem doing that either

**Comment by GuSilverFlame** [comment]
@arthurra I changed the card style to be a little more like the account card from #744 
is it alright like that?

File: apps/explorer/lib/explorer/chain/token.ex (hunk 12 above)
Comment (igorffs) [reply apps/explorer/lib/explorer/chain/token.ex]: Oh, I see! 

https://github.com/poanetwork/blockscout/blob/4f9e7a5963c5f983ba9b7a8830adddb5cdb91420/apps/explorer/lib/explorer/chain.ex#L1673-L1677

The `token_with_number_of_transfers_from_address` function is basically using this one here. 🤔 I think we should move the tests closer to where the implementation is actually being done. wdyt @amandasposito @Lucasnar @feliperenan?

Comment (amandasposito) [reply apps/explorer/lib/explorer/chain/token.ex]: I'm okay with leaving the tests at the `chain_test.exs` 👍

File: apps/explorer/lib/explorer/chain/address/token_card.ex [hunk 15]
```diff
@@ -0,0 +1,95 @@
+defmodule Explorer.Chain.Address.TokenCard do
```
Comment (amandasposito) [inline apps/explorer/lib/explorer/chain/address/token_card.ex]: This module is not much expressive about its intention and what it's responsibility. What is a `card`? Why does a `card` need to know about `Token` and `TokenBalance`?

It is not clear what problem it’s solving.

The problem we are trying to solve, it is to have the security, at the `web` level, to pattern match against a token struct so we can properly work with the `paging params`. This is a viewer problem and it shouldn't reflect on our domain like that.

I believe the best option here is to revert [this](https://github.com/poanetwork/blockscout/commit/66a2dd7) specific change and include only a struct we can use in our query, something like https://github.com/poanetwork/blockscout/blob/master/apps/explorer/lib/explorer/paging_options.ex.

WDYT? cc @igorffs

Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/address/token_card.ex]: the naming is probably way off because I'm horrible naming things in a comprehensible way but making a module for this and separating the responsibility of this query from Token was something igorffs had discussed with me
the struct itself was more of an afterthought, something like "hey, I can fit a struct here and solve two things at once, nice!"

Comment (igorffs) [reply apps/explorer/lib/explorer/chain/address/token_card.ex]: Another thing that drove us to this was thinking that this would probably be a good point for considering a denormalization because of the growing complexity to create this projection here (all the joins, distinct and other aspects present in our queries), which would probably lead for a similar modeling.

Comment (amandasposito) [reply apps/explorer/lib/explorer/chain/address/token_card.ex]: IMO this is not a Read Model, it doesn't have independent read and write models, it is just confusing because you don't know what this module here does. The name problem we were having with the function, it is present here in the module with a name that is also not clear.

File: apps/explorer/lib/explorer/chain/address/token.ex [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+  along with the latest balance of each token for that address and the count of
+  transfers the address had with each token
+
+  warning: tokens with a latest balance of zero will be ignored
+
+  In order to fetch a token, the given address must have transfered tokens to or received tokens
+  from another address. This quey orders by the token type and name.
+  """
+  @spec for_address_and_page(Hash.t(), [paging_options()]) :: %Ecto.Query{}
+  def for_address_and_page(address_hash, options \\ []) do
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/address/token.ex]: I think we can change this here to "list_address_tokens", "all" or something that express that we want a list of `Address.Token`s. wdyt?

File: apps/explorer/lib/explorer/chain/address/token.ex [hunk 17]
```diff
@@ -0,0 +1,96 @@
+defmodule Explorer.Chain.Address.Token do
+  @moduledoc """
+  an abstraction to gather all data of a token in a specific address
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/address/token.ex]:
```
A projection that represents the relation between a Token and a specific Address.

This representation is expressed by the following attributes:

- contract_address_hash - A Hash, Address of a Token contract.
- name - a Token name.
- symbol - a  Token symbol.
- type - a Token type.
- balance - how much tokens (TokenBalance) an Address has from a Token.
- transfer_count - a count of how many TokenTransfers of a Token an Address was involved.
```

@GuSilverFlame ☝️ wdyt?

File: apps/explorer/lib/explorer/chain/address/token.ex [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+  import Ecto.Query
+  alias Explorer.PagingOptions
+  alias Explorer.Chain
+
+  alias Explorer.Chain.{Hash, Address, Address.TokenBalance}
+
+  @default_paging_options %PagingOptions{page_size: 50}
+  @typep paging_options :: {:paging_options, PagingOptions.t()}
+
+  @enforce_keys [:contract_address_hash, :inserted_at, :name, :symbol, :value, :decimals, :type, :transfers]
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/address/token.ex]: It'd nice if we:
- use balance instead of value, because it's a name used by the TokenBalance and it'd be easier to relate them.
- and transfers_count instead of just transfer, because we're getting only the count. \o

File: apps/explorer/test/explorer/chain/address/token_test.exs [hunk 19]
```diff
@@ -0,0 +1,31 @@
+defmodule Explorer.Chain.Address.TokenTest do
+  use Explorer.DataCase
+
+  alias Explorer.Chain.Address
+  alias Explorer.Chain.Token
+  alias Explorer.PagingOptions
+
```
Comment (igorffs) [inline apps/explorer/test/explorer/chain/address/token_test.exs]: Let's move the tests that are in the `chain_test.ex` related to the old function to this file, by testing the query function here. 👍

File: apps/explorer/lib/explorer/chain/address/token.ex (hunk 17 above)
Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/address/token.ex]: looks nice!

File: apps/explorer/lib/explorer/chain.ex [hunk 20]
```diff
@@ -1700,35 +1700,10 @@ defmodule Explorer.Chain do
   @spec tokens_with_number_of_transfers_from_address(Hash.Address.t(), [any()]) :: []
   def tokens_with_number_of_transfers_from_address(address_hash, paging_options \\ []) do
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain.ex]: This public API name also needs to change, maybe: "address_tokens_with_balance".

File: apps/explorer/test/explorer/chain/address/token_test.exs [hunk 21]
```diff
@@ -0,0 +1,263 @@
+defmodule Explorer.Chain.Address.TokenTest do
+  use Explorer.DataCase
+
+  alias Explorer.Repo
+  alias Explorer.Chain.Address
+  alias Explorer.Chain.Token
+  alias Explorer.PagingOptions
+
+  describe "list_address_tokens_with_balance/2" do
```
Comment (igorffs) [inline apps/explorer/test/explorer/chain/address/token_test.exs]: Can a specific test for the ordering rules?

File: apps/explorer/lib/explorer/chain/address/token.ex [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+  import Ecto.Query
+  alias Explorer.{PagingOptions, Chain}
+
+  alias Explorer.Chain.{Hash, Address, Address.TokenBalance}
+
+  @default_paging_options %PagingOptions{page_size: 50}
+  @typep paging_options :: {:paging_options, PagingOptions.t()}
+
+  @enforce_keys [:contract_address_hash, :inserted_at, :name, :symbol, :balance, :decimals, :type, :transfers_count]
+  defstruct @enforce_keys
```
Comment (feliperenan) [inline apps/explorer/lib/explorer/chain/address/token.ex]: 🎨 We could move it to be right below in the `@moduledoc`. It also would fit with that documentation :).

File: apps/explorer/lib/explorer/chain/address/token.ex [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+    from(
+      t in queryable,
+      join: tb in subquery(last_balance_query),
+      on: tb.token_contract_address_hash == t.contract_address_hash
+    )
+  end
+
+  @doc """
+  adds to the passed `Ecto.Query` a criteria indicating the first token of the current page
```
Comment (feliperenan) [inline apps/explorer/lib/explorer/chain/address/token.ex]: We could improve this documentation with something like that:

```Elixir
@doc """
Builds the pagination according to the given key within `PagingOptions`.

* it just returns the given query when the key is nil.
* it composes another where clause considering the `type`, `name` and `inserted_at`.

"""
```

Wdyt?

Comment (GuSilverFlame) [reply apps/explorer/lib/explorer/chain/address/token.ex]: seems good to me

**Comment by GuSilverFlame** [comment]
@acravenho are we good to move on with this one?

**Comment by acravenho** [comment]
@GuSilverFlame I'm going to take another quick look.
